library/test/src/lib.rs

```diff
@@ -299,6 +299,30 @@ where
         })
     }
 
+    if !opts.nocapture {
+        // If we encounter a non-unwinding panic, flush any captured output from the current test,
+        // and stop  capturing output to ensure that the non-unwinding panic message is visible.
+        // We also acquire the locks for both output streams to prevent output from other threads
+        // from interleaving with the panic message or appearing after it.
+        let builtin_panic_hook = panic::take_hook();
+        let hook = Box::new({
+            move |info: &'_ PanicInfo<'_>| {
+                if !info.can_unwind() {
+                    std::mem::forget(std::io::stderr().lock());
+                    let mut stdout = ManuallyDrop::new(std::io::stdout().lock());
+                    if let Some(captured) = io::set_output_capture(None) {
+                        if let Ok(data) = captured.lock() {
+                            let _ = stdout.write_all(&data);
+                            let _ = stdout.flush();
+                        }
+                    }
+                }
+                builtin_panic_hook(info);
+            }
+        });
+        panic::set_hook(hook);
```

Comment:
Should this function reset the hook before returning? If not it might be worth at least documenting in its doc comment that it clutters the panic hook and basically assumes the process exits after this function returns.

Reply:
I shifted this code out to `test_main`, which calls `process::exit` along error paths. I'm not sure of an ergonomic way to swap the panic hook back to the previous setting, but I think putting it in `test_main` makes it unlikely anyone will try to run code after the hook is changed.

Reply:
Agreed.